src/core/utils/global_address.cpp

```diff
@@ -0,0 +1,264 @@
+                }
+
+                otAddUnicastAddress(aInstance, address);
+                break;
+            }
+        }
+    }
+}
+
+ThreadError Slaac::createRandomIid(otNetifAddress *aAddress, void *aContext)
+{
+    (void) aContext;
+
+    for (size_t i = 8; i < sizeof(aAddress[i].mAddress); i++)
+    {
+        aAddress->mAddress.mFields.m8[i] = (uint8_t)otPlatRandomGet();
+    }
+
+    return kThreadError_None;
+}
+
+ThreadError SemanticallyOpaqueIidGenerator::createIid(otNetifAddress *aAddress, void *aContext)
+{
+    ThreadError error = kThreadError_None;
+    otSemanticallyOpaqueIidGeneratorData *data = (otSemanticallyOpaqueIidGeneratorData *)aContext;
+
+    VerifyOrExit(aContext != NULL, error = kThreadError_InvalidArgs);
+
+    for (uint32_t i = 0; i <= 255; i++)
+    {
+        error = createIid(aAddress, data);
+        VerifyOrExit(error == kThreadError_Failed,);
+
+        data->mDadCounter++;
+    }
+
+exit:
+    return error;
+}
+
+ThreadError SemanticallyOpaqueIidGenerator::createIid(otNetifAddress *aAddress,
+                                                      otSemanticallyOpaqueIidGeneratorData *aData)
+{
+    ThreadError error = kThreadError_None;
+    Crypto::Sha256 sha256;
+    uint8_t hash[Crypto::Sha256::kHashSize];
+
+    sha256.Start();
+
+    sha256.Update(aAddress->mAddress.mFields.m8, aAddress->mPrefixLength / 8);
+
+    VerifyOrExit(aData->mInterfaceId != NULL, error = kThreadError_InvalidArgs);
+    sha256.Update(aData->mInterfaceId, aData->mInterfaceIdLength);
+
+    if (aData->mNetworkIdLength)
+    {
+        VerifyOrExit(aData->mNetworkId != NULL, error = kThreadError_InvalidArgs);
+        sha256.Update(aData->mNetworkId, aData->mNetworkIdLength);
+    }
+
+    sha256.Update((uint8_t *) &aData->mDadCounter, sizeof(aData->mDadCounter));
+
+    VerifyOrExit(aData->mSecretKey != NULL, error = kThreadError_InvalidArgs);
+    sha256.Update(aData->mSecretKey, aData->mSecretKeyLength);
+
+    sha256.Finish(hash);
+
+    memcpy(&aAddress->mAddress.mFields.m8[8], &hash[sizeof(hash) - 8], 8);
+
+    VerifyOrExit(!IsAddressRegistered(aAddress, aData->mAddresses), error = kThreadError_Failed);
+    VerifyOrExit(!IsAddressReserved(aAddress), error = kThreadError_Failed);
+
+exit:
+    return error;
+}
+
+bool SemanticallyOpaqueIidGenerator::IsAddressRegistered(otNetifAddress *aCreatedAddress, otNetifAddress *aAddressList)
+{
+    otNetifAddress *address = aAddressList;
+
+    while (address != NULL)
+    {
+        if (0 == memcmp(aCreatedAddress->mAddress.mFields.m8, address->mAddress.mFields.m8,
+                        sizeof(address->mAddress.mFields)))
+        {
+            return true;
+        }
+
+        address = address->mNext;
+    }
+
+    return false;
+}
+
+bool SemanticallyOpaqueIidGenerator::IsAddressReserved(otNetifAddress *aCreatedAddress)
+{
+    // Subnet-Router Anycast
+    if ((aCreatedAddress->mAddress.mFields.m32[2] == 0) &&
+        (aCreatedAddress->mAddress.mFields.m32[3] == 0))
```

Comment:
`Ip6::Address` includes a bunch of helper methods to check address types.  Shall we add `IsSubnetRouterAnycast()` and `IsSubnetAnycast()` to `Ip6::Address`?